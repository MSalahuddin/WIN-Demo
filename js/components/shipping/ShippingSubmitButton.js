import React from 'react';
import styled from 'styled-components/native';
import PropTypes from 'prop-types';
import { Platform } from 'react-native';
import Text, { SIZE, FONT_FAMILY } from '../common/Text';
import Button, { BUTTON_COLOR_SCHEME } from '../common/Button';
import { scaleHeight } from '../../platformUtils';
import { color } from '../../styles';
import { shippingStrings } from '../../stringConstants';

const ButtonWrapper = styled(Button)`
  margin-top: ${scaleHeight(32)};
`;

const Container = styled.View`
  align-items: center;
  margin-bottom: ${scaleHeight(74)};
`;

const TextContainer = styled.View`
  align-items: center;
  justify-content: center;
  margin-top: ${scaleHeight(30)};
`;

const ValidationContainer = styled.View`
  align-items: center;
  justify-content: center;
`;

const TextWrapper = styled(Text)`
  color: ${color.bluishColor};
  margin-top: ${Platform.OS === 'android' ? scaleHeight(0) : scaleHeight(8)};
`;

const ShippingSubmitButton = ({ fieldsNeedValidate, error, onPress, touched, disabledButton }) => {
  const isFormUntoched = fieldsNeedValidate.length === 8;
  // form is complete when every field is prepopulated or
  // at least one field is touched with no error occurs
  const isFormComplete =
    fieldsNeedValidate.length === 0 || (Object.keys(touched).length >= 1 && Object.keys(error).length === 0);
  const fieldsTouchedHasError = Object.keys(error).filter(value => Object.keys(touched).includes(value));
  const isFormErrored = fieldsTouchedHasError.length > 0;

  const validationSummary = Object.keys(error).map(x => 
  (<Text
    fontFamily={FONT_FAMILY.BOLD}
    size={SIZE.SMALL}
    color={color.warningRed}
  >
  {error[x]}
  </Text>))

  const bannerMessage = () => {
    switch (true) {
      case isFormUntoched:
        return shippingStrings.fillOutForm;
      case isFormErrored:
        return shippingStrings.oopsCorrectErrors;
      default:
        return shippingStrings.yourPrizeAwaits;
    }
  };
  const theme = () => {
    if (disabledButton || isFormErrored) {
      return BUTTON_COLOR_SCHEME.GREY
    }
    if (isFormComplete) {
      return BUTTON_COLOR_SCHEME.WHITE
    }
    return BUTTON_COLOR_SCHEME.WHITE
  }

  const disabled = () => {
    if (disabledButton) {
      return disabledButton
    }
    return !isFormComplete || isFormErrored
  }

  return (
    <Container>
      <ButtonWrapper
        testID="shipping-button"
        height={88}
        width={271}
        borderRadius={44}
        theme={theme()}
        disabled={disabled()}
        onPress={onPress}
      >
        <TextWrapper size={SIZE.XLARGE}>{shippingStrings.submitShip}</TextWrapper>
      </ButtonWrapper>
      <TextContainer>
        <Text
          fontFamily={FONT_FAMILY.BOLD}
          size={SIZE.SMALL}
          color={isFormErrored ? color.warningRed : color.white}
        >
          {bannerMessage()}
        </Text>
      </TextContainer>
      <ValidationContainer>
          {validationSummary}
      </ValidationContainer>
    </Container>
  );
};

ShippingSubmitButton.propTypes = {
  error: PropTypes.shape({}).isRequired,
  onPress: PropTypes.func.isRequired,
  touched: PropTypes.shape({}).isRequired,
  fieldsNeedValidate: PropTypes.arrayOf(PropTypes.string).isRequired,
  disabledButton: PropTypes.func.isRequired
};

export default ShippingSubmitButton;
