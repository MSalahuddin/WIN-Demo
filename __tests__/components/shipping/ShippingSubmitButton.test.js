import React from 'react';
import { render, fireEvent } from 'react-native-testing-library';

import ShippingSubmitButton from '../../../js/components/shipping/ShippingSubmitButton';
import MockProvider from '../../MockProvider';

describe('Component Button', () => {
  const onPress = jest.fn();
  const touched = {
    address1: true,
    city: true,
    zipCode: true,
    state: true,
    firstName: true,
    lastName: true,
    email: true,
    phoneNumber: true
  };

  const fieldsNeedValidate = ['address1', 'city', 'email', 'firstName', 'lastName', 'phoneNumber', 'state', 'zipCode'];

  test('renders disabled button correctly', () => {
    const output = render(
      <MockProvider>
        <ShippingSubmitButton onPress={onPress} error={{}} touched={{}} fieldsNeedValidate={fieldsNeedValidate} />
      </MockProvider>
    );
    expect(output).toMatchSnapshot();
  });

  test('renders button correctly', () => {
    const output = render(
      <MockProvider>
        <ShippingSubmitButton onPress={onPress} error={{}} touched={touched} fieldsNeedValidate={[]} />
      </MockProvider>
    );
    expect(output).toMatchSnapshot();
  });

  test('calls onPress when button is pressed', () => {
    const { getByTestId } = render(
      <MockProvider>
        <ShippingSubmitButton onPress={onPress} error={{}} touched={touched} fieldsNeedValidate={fieldsNeedValidate} />
      </MockProvider>
    );
    fireEvent(getByTestId('shipping-button'), 'press');
    expect(onPress).toHaveBeenCalled();
  });
});
