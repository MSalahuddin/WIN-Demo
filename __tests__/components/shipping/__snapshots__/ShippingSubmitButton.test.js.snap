// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Button renders button correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "marginBottom": 148,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginHorizontal": 10,
          "marginVertical": 10,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(116, 158, 253, 1)",
          "borderRadius": 88,
          "height": 176,
          "justifyContent": "center",
          "marginTop": 64,
          "opacity": 1,
          "width": 542,
        }
      }
      testID="shipping-button"
    >
      <View
        borderRadius={44}
        colorScheme="rgba(165, 193, 255, 1)"
        height={88}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(165, 193, 255, 1)",
              "borderRadius": 88,
              "height": 166,
              "justifyContent": "center",
              "width": 536,
            },
          ]
        }
        width={271}
      >
        <View
          borderRadius={44}
          colorScheme="rgba(127, 166, 253, 1)"
          height={88}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "rgba(127, 166, 253, 1)",
                "borderRadius": 108,
                "height": 162,
                "justifyContent": "center",
                "marginTop": 8,
                "width": 536,
              },
            ]
          }
          width={271}
        >
          <Text
            alignCenter={false}
            allowFontScaling={true}
            color="#000000"
            fontFamily="MEDIUM"
            isUnderlined={false}
            isUppercase={false}
            maxFontSizeMultiplier={1.3}
            size="XLARGE"
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "Calibre-Medium",
                  "fontSize": 36,
                  "justifyContent": "center",
                  "textAlign": "left",
                  "textDecorationColor": "#000000",
                  "textDecorationLine": "none",
                  "textDecorationStyle": "solid",
                  "textTransform": "none",
                },
                Object {
                  "color": "#FFFFFF",
                  "marginTop": 16,
                },
              ]
            }
          >
            Submit and ship
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 8,
          "justifyContent": "center",
          "paddingBottom": 38,
          "paddingTop": 114,
          "position": "absolute",
          "top": 152,
          "width": 638,
          "zIndex": -1,
        },
      ]
    }
  >
    <Text
      alignCenter={false}
      allowFontScaling={true}
      color="#222222"
      fontFamily="REGULAR"
      isUnderlined={false}
      isUppercase={false}
      maxFontSizeMultiplier={1.3}
      size="XSMALL"
      style={
        Array [
          Object {
            "color": "#222222",
            "fontFamily": "CalibreRegular",
            "fontSize": 24,
            "justifyContent": "center",
            "textAlign": "left",
            "textDecorationColor": "#222222",
            "textDecorationLine": "none",
            "textDecorationStyle": "solid",
            "textTransform": "none",
          },
        ]
      }
    >
      Your prize awaits!
    </Text>
  </View>
</View>
`;

exports[`Component Button renders disabled button correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "marginBottom": 148,
      },
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginHorizontal": 10,
          "marginVertical": 10,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(160,160,160,1)",
          "borderRadius": 88,
          "height": 176,
          "justifyContent": "center",
          "marginTop": 64,
          "opacity": 1,
          "width": 542,
        }
      }
      testID="shipping-button"
    >
      <View
        borderRadius={44}
        colorScheme="rgba(209,209,209,1)"
        height={88}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(209,209,209,1)",
              "borderRadius": 88,
              "height": 166,
              "justifyContent": "center",
              "width": 536,
            },
          ]
        }
        width={271}
      >
        <View
          borderRadius={44}
          colorScheme="rgba(183,183,183,1)"
          height={88}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "rgba(183,183,183,1)",
                "borderRadius": 108,
                "height": 162,
                "justifyContent": "center",
                "marginTop": 8,
                "width": 536,
              },
            ]
          }
          width={271}
        >
          <Text
            alignCenter={false}
            allowFontScaling={true}
            color="#000000"
            fontFamily="MEDIUM"
            isUnderlined={false}
            isUppercase={false}
            maxFontSizeMultiplier={1.3}
            size="XLARGE"
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "Calibre-Medium",
                  "fontSize": 36,
                  "justifyContent": "center",
                  "textAlign": "left",
                  "textDecorationColor": "#000000",
                  "textDecorationLine": "none",
                  "textDecorationStyle": "solid",
                  "textTransform": "none",
                },
                Object {
                  "color": "#FFFFFF",
                  "marginTop": 16,
                },
              ]
            }
          >
            Submit and ship
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 8,
          "justifyContent": "center",
          "paddingBottom": 38,
          "paddingTop": 114,
          "position": "absolute",
          "top": 152,
          "width": 638,
          "zIndex": -1,
        },
      ]
    }
  >
    <Text
      alignCenter={false}
      allowFontScaling={true}
      color="#222222"
      fontFamily="REGULAR"
      isUnderlined={false}
      isUppercase={false}
      maxFontSizeMultiplier={1.3}
      size="XSMALL"
      style={
        Array [
          Object {
            "color": "#222222",
            "fontFamily": "CalibreRegular",
            "fontSize": 24,
            "justifyContent": "center",
            "textAlign": "left",
            "textDecorationColor": "#222222",
            "textDecorationLine": "none",
            "textDecorationStyle": "solid",
            "textTransform": "none",
          },
        ]
      }
    >
      Fill out the form to ship your prize!
    </Text>
  </View>
</View>
`;
